lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3371,9 +3378,44 @@ namespace {
               decl);
         }
       }
+    }
 
-      return importFunctionDecl(decl, importedName, correctSwiftName,
-                                std::nullopt);
+    /// Emit warnings for unannotated C++ APIs returning FRTs
+    void
+    diagnoseUnannotatedCxxApisReturningFRT(const clang::FunctionDecl *decl) {
```

Comment:
Suggestion: focus on checking whether this `decl` needs an annotation, and defer diagnosis to the caller:

```suggestion
    bool needsFRTReturnAttr(const clang::FunctionDecl *decl) {
```

Then you can simply `return true` or `return false` in this function, rather than accumulating a boolean variable.

Then the caller will change from:

```cpp
        } else if (!returnsRetainedAttrIsPresent &&
                    !returnsUnretainedAttrIsPresent) {
           diagnoseUnannotatedCxxApisReturningFRT(decl);
        }
```

to:

```cpp
        } else if (!returnsRetainedAttrIsPresent &&
                    !returnsUnretainedAttrIsPresent &&
                    needsFRTReturnAttr(decl)) {
           Impl.diagnose(loc, diag::no_returns_retained_returns_unretained, decl);
        }
```

(modulo `clang-format`)

This structure keeps all `Impl.diagnose()` calls in the same place, and
a bonus is that you get to reuse the `loc` variable you constructed earlier.